File: core/src/androidTest/java/org/hisp/dhis/android/core/event/internal/EventAPIRealShould.java
/*
 *  Copyright (c) 2004-2022, University of Oslo
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *  Redistributions of source code must retain the above copyright notice, this
 *  list of conditions and the following disclaimer.
 *
 *  Redistributions in binary form must reproduce the above copyright notice,
 *  this list of conditions and the following disclaimer in the documentation
 *  and/or other materials provided with the distribution.
 *  Neither the name of the HISP project nor the names of its contributors may
 *  be used to endorse or promote products derived from this software without
 *  specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 *  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 *  ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 *  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.hisp.dhis.android.core.event.internal;

import static com.google.common.truth.Truth.assertThat;
import static org.hisp.dhis.android.core.event.internal.EventUtils.assertEvent;
import static org.hisp.dhis.android.core.event.internal.EventUtils.createEventWithFutureDate;
import static org.hisp.dhis.android.core.event.internal.EventUtils.createEventWithInvalidAttributeOptionCombo;
import static org.hisp.dhis.android.core.event.internal.EventUtils.createEventWithInvalidDataValues;
import static org.hisp.dhis.android.core.event.internal.EventUtils.createEventWithInvalidOrgunit;
import static org.hisp.dhis.android.core.event.internal.EventUtils.createEventWithInvalidProgram;
import static org.hisp.dhis.android.core.event.internal.EventUtils.createValidEvent;
import static org.hisp.dhis.android.core.imports.ImportStatus.ERROR;
import static org.hisp.dhis.android.core.imports.ImportStatus.SUCCESS;
import static org.hisp.dhis.android.core.imports.ImportStatus.WARNING;

import org.hisp.dhis.android.core.BaseRealIntegrationTest;
import org.hisp.dhis.android.core.arch.api.executors.internal.APICallExecutor;
import org.hisp.dhis.android.core.arch.api.executors.internal.APICallExecutorImpl;
import org.hisp.dhis.android.core.event.Event;
import org.hisp.dhis.android.core.imports.internal.EventImportSummary;
import org.hisp.dhis.android.core.imports.internal.EventWebResponse;
import org.hisp.dhis.android.core.maintenance.D2Error;
import org.hisp.dhis.android.core.organisationunit.OrganisationUnitMode;
import org.junit.Assert;
import org.junit.Before;

import java.util.Arrays;
import java.util.Collections;

public abstract class EventAPIRealShould extends BaseRealIntegrationTest {

    // API version dependant parameters
    private String serverUrl;
    private String strategy;
    private String ouMode = OrganisationUnitMode.ACCESSIBLE.name();

    private APICallExecutor apiCallExecutor;

    private EventService eventService;

    EventAPIRealShould(String serverUrl, String strategy) {
        super();
        this.serverUrl = serverUrl;
        this.strategy = strategy;
    }

    @Before
    public void setUp() {
        super.setUp();
        apiCallExecutor = APICallExecutorImpl.create(d2.databaseAdapter(), null);

        eventService = d2.retrofit().create(EventService.class);
    }

    //@Test
    public void valid_events() throws Exception {
        login();

        Event validEvent1 = createValidEvent();

        Event validEvent2 = createValidEvent();

        EventPayload payload = new EventPayload();
        payload.events = Arrays.asList(validEvent1, validEvent2);

        EventWebResponse response = apiCallExecutor.executeObjectCallWithAcceptedErrorCodes(eventService
                .postEvents(payload, this.strategy), Collections.singletonList(409), EventWebResponse.class);

        assertThat(response.response().status()).isEqualTo(SUCCESS);

        for (EventImportSummary importSummary : response.response().importSummaries()) {
            if (validEvent1.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
            else if (validEvent2.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
        }

        // Check server status
        Event serverValidEvent1 = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent1.uid(), EventFields.allFields, ouMode));
        Event serverValidEvent2 = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent2.uid(), EventFields.allFields, ouMode));

        assertThat(serverValidEvent1).isNotNull();
        assertThat(serverValidEvent2).isNotNull();
    }

    //@Test
    public void event_with_invalid_orgunit() throws Exception {
        login();

        Event validEvent = createValidEvent();

        Event invalidEvent = createEventWithInvalidOrgunit();

        EventPayload payload = new EventPayload();
        payload.events = Arrays.asList(validEvent, invalidEvent);

        EventWebResponse response = apiCallExecutor.executeObjectCallWithAcceptedErrorCodes(eventService
                .postEvents(payload, this.strategy), Collections.singletonList(409), EventWebResponse.class);

        assertThat(response.response().status()).isEqualTo(ERROR);

        for (EventImportSummary importSummary : response.response().importSummaries()) {
            if (validEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
            else if (invalidEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, ERROR);
            }
        }

        // Check server status
        Event serverValidEvent = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent.uid(), EventFields.allFields, ouMode));

        assertThat(serverValidEvent).isNotNull();

        try {
            apiCallExecutor.executeObjectCall(eventService.getEvent(invalidEvent.uid(), EventFields.allFields, ouMode));
            Assert.fail("Should not reach that line");
        } catch (D2Error e) {
            assertThat(e.httpErrorCode()).isEqualTo(404);
        }
    }

    //@Test
    public void event_with_invalid_attribute_option_combo() throws Exception {
        login();

        Event validEvent = createValidEvent();

        Event invalidEvent = createEventWithInvalidAttributeOptionCombo();

        EventPayload payload = new EventPayload();
        payload.events = Arrays.asList(validEvent, invalidEvent);

        EventWebResponse response = apiCallExecutor.executeObjectCallWithAcceptedErrorCodes(eventService
                .postEvents(payload, this.strategy), Collections.singletonList(409), EventWebResponse.class);

        assertThat(response.response().status()).isEqualTo(ERROR);

        for (EventImportSummary importSummary : response.response().importSummaries()) {
            if (validEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
            else if (invalidEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, ERROR);
            }
        }

        // Check server status
        Event serverValidEvent = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent.uid(), EventFields.allFields, ouMode));

        assertThat(serverValidEvent).isNotNull();

        try {
            apiCallExecutor.executeObjectCall(eventService.getEvent(invalidEvent.uid(), EventFields.allFields, ouMode));
            Assert.fail("Should not reach that line");
        } catch (D2Error e) {
            assertThat(e.httpErrorCode()).isEqualTo(404);
        }
    }

    //@Test
    public void event_with_future_date_does_not_fail() throws Exception {
        login();

        Event validEvent1 = createValidEvent();

        Event validEvent2 = createEventWithFutureDate();

        EventPayload payload = new EventPayload();
        payload.events = Arrays.asList(validEvent1, validEvent2);

        EventWebResponse response = apiCallExecutor.executeObjectCallWithAcceptedErrorCodes(eventService
                .postEvents(payload, this.strategy), Collections.singletonList(409), EventWebResponse.class);

        assertThat(response.response().status()).isEqualTo(SUCCESS);

        for (EventImportSummary importSummary : response.response().importSummaries()) {
            if (validEvent1.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
            else if (validEvent2.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
        }

        // Check server status
        Event serverValidEvent1 = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent1.uid(), EventFields.allFields, ouMode));
        Event serverValidEvent2 = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent2.uid(), EventFields.allFields, ouMode));

        assertThat(serverValidEvent1).isNotNull();
        assertThat(serverValidEvent2).isNotNull();
    }

    //@Test
    public void event_with_invalid_program() throws Exception {
        login();

        Event validEvent = createValidEvent();

        Event invalidEvent = createEventWithInvalidProgram();

        EventPayload payload = new EventPayload();
        payload.events = Arrays.asList(validEvent, invalidEvent);

        EventWebResponse response = apiCallExecutor.executeObjectCallWithAcceptedErrorCodes(eventService
                .postEvents(payload, this.strategy), Collections.singletonList(409), EventWebResponse.class);

        assertThat(response.response().status()).isEqualTo(ERROR);

        for (EventImportSummary importSummary : response.response().importSummaries()) {
            if (validEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
            else if (invalidEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, ERROR);
            }
        }

        // Check server status
        Event serverValidEvent = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent.uid(), EventFields.allFields, ouMode));

        assertThat(serverValidEvent).isNotNull();

        try {
            apiCallExecutor.executeObjectCall(eventService.getEvent(invalidEvent.uid(), EventFields.allFields, ouMode));
            Assert.fail("Should not reach that line");
        } catch (D2Error e) {
            assertThat(e.httpErrorCode()).isEqualTo(404);
        }
    }

    //@Test
    public void event_with_invalid_data_values() throws Exception {
        login();

        Event validEvent = createValidEvent();

        Event invalidEvent = createEventWithInvalidDataValues();

        EventPayload payload = new EventPayload();
        payload.events = Arrays.asList(validEvent, invalidEvent);

        EventWebResponse response = apiCallExecutor.executeObjectCallWithAcceptedErrorCodes(eventService
                .postEvents(payload, this.strategy), Collections.singletonList(409), EventWebResponse.class);

        assertThat(response.response().status()).isEqualTo(WARNING);

        for (EventImportSummary importSummary : response.response().importSummaries()) {
            if (validEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, SUCCESS);
            }
            else if (invalidEvent.uid().equals(importSummary.reference())) {
                assertEvent(importSummary, WARNING);
            }
        }

        // Check server status
        Event serverValidEvent = apiCallExecutor.executeObjectCall(eventService.getEvent(validEvent.uid(),
                EventFields.allFields, ouMode));
        Event serverInvalidEvent = apiCallExecutor.executeObjectCall(eventService.getEvent(invalidEvent.uid(),
                EventFields.allFields, ouMode));

        assertThat(serverValidEvent).isNotNull();
        assertThat(serverValidEvent.trackedEntityDataValues().size()).isEqualTo(2);

        assertThat(serverInvalidEvent).isNotNull();
        assertThat(serverInvalidEvent.trackedEntityDataValues()).isNull();
    }

    private void login() {
        d2.userModule().logIn(username, password, url).blockingGet();
    }
}

File: core/src/androidTest/java/org/hisp/dhis/android/core/event/internal/EventAPIRealShould.kt
/*
 *  Copyright (c) 2004-2023, University of Oslo
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *  Redistributions of source code must retain the above copyright notice, this
 *  list of conditions and the following disclaimer.
 *
 *  Redistributions in binary form must reproduce the above copyright notice,
 *  this list of conditions and the following disclaimer in the documentation
 *  and/or other materials provided with the distribution.
 *  Neither the name of the HISP project nor the names of its contributors may
 *  be used to endorse or promote products derived from this software without
 *  specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 *  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 *  ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 *  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.hisp.dhis.android.core.event.internal

import com.google.common.truth.Truth.assertThat
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import org.hisp.dhis.android.core.BaseRealIntegrationTest
import org.hisp.dhis.android.core.arch.api.executors.internal.CoroutineAPICallExecutor
import org.hisp.dhis.android.core.arch.helpers.Result
import org.hisp.dhis.android.core.imports.ImportStatus
import org.hisp.dhis.android.core.imports.internal.EventWebResponse
import org.hisp.dhis.android.core.maintenance.D2Error
import org.hisp.dhis.android.core.organisationunit.OrganisationUnitMode
import org.junit.Assert
import org.junit.Before
import java.util.*

@OptIn(ExperimentalCoroutinesApi::class)
abstract class EventAPIRealShould internal constructor(
    // API version dependant parameters
    private val serverUrl: String,
    private val strategy: String,
) : BaseRealIntegrationTest() {
    private val ouMode = OrganisationUnitMode.ACCESSIBLE.name
    private lateinit var apiCallExecutor: CoroutineAPICallExecutor
    private lateinit var eventService: EventService

    @Before
    override fun setUp() {
        super.setUp()
        apiCallExecutor = d2.coroutineAPICallExecutor()
        eventService = EventService(d2.httpServiceClient())
    }

    // @Test
    @Throws(Exception::class)
    fun valid_events() = runTest {
        login()
        val validEvent1 = EventUtils.createValidEvent()
        val validEvent2 = EventUtils.createValidEvent()
        val payload = EventPayload()
        payload.events = listOf(validEvent1, validEvent2)

        val response = postPayload(payload).getOrThrow()

        assertThat(response.response()!!.status()).isEqualTo(ImportStatus.SUCCESS)

        for (importSummary in response.response()!!.importSummaries()!!) {
            if (validEvent1.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            } else if (validEvent2.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            }
        }

        // Check server status
        val serverValidEvent1 = apiCallExecutor.wrap {
            eventService.getEvent(validEvent1.uid(), EventFields.allFields, ouMode)
        }.getOrThrow()

        val serverValidEvent2 = apiCallExecutor.wrap {
            eventService.getEvent(validEvent2.uid(), EventFields.allFields, ouMode)
        }.getOrThrow()

        assertThat(serverValidEvent1).isNotNull()
        assertThat(serverValidEvent2).isNotNull()
    }

    // @Test
    @Throws(Exception::class)
    fun event_with_invalid_orgunit() = runTest {
        login()
        val validEvent = EventUtils.createValidEvent()
        val invalidEvent = EventUtils.createEventWithInvalidOrgunit()
        val payload = EventPayload()
        payload.events = listOf(validEvent, invalidEvent)

        val response = postPayload(payload).getOrThrow()

        assertThat(response.response()!!.status()).isEqualTo(ImportStatus.ERROR)

        for (importSummary in response.response()!!.importSummaries()!!) {
            if (validEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            } else if (invalidEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.ERROR)
            }
        }

        // Check server status
        val serverValidEvent = apiCallExecutor.wrap {
            eventService.getEvent(validEvent.uid(), EventFields.allFields, ouMode)
        }.getOrThrow()

        assertThat(serverValidEvent).isNotNull()

        try {
            apiCallExecutor.wrap {
                eventService.getEvent(
                    invalidEvent.uid(),
                    EventFields.allFields,
                    ouMode,
                )
            }.getOrThrow()
            Assert.fail("Should not reach that line")
        } catch (e: D2Error) {
            assertThat(e.httpErrorCode()).isEqualTo(404)
        }
    }

    // @Test
    @Throws(Exception::class)
    fun event_with_invalid_attribute_option_combo() = runTest {
        login()
        val validEvent = EventUtils.createValidEvent()
        val invalidEvent = EventUtils.createEventWithInvalidAttributeOptionCombo()
        val payload = EventPayload()
        payload.events = listOf(validEvent, invalidEvent)

        val response = postPayload(payload).getOrThrow()

        assertThat(response.response()!!.status()).isEqualTo(ImportStatus.ERROR)

        for (importSummary in response.response()!!.importSummaries()!!) {
            if (validEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            } else if (invalidEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.ERROR)
            }
        }

        // Check server status
        val serverValidEvent = apiCallExecutor.wrap {
            eventService.getEvent(validEvent.uid(), EventFields.allFields, ouMode)
        }.getOrThrow()

        assertThat(serverValidEvent).isNotNull()

        try {
            apiCallExecutor.wrap {
                eventService.getEvent(
                    invalidEvent.uid(),
                    EventFields.allFields,
                    ouMode,
                )
            }.getOrThrow()
            Assert.fail("Should not reach that line")
        } catch (e: D2Error) {
            assertThat(e.httpErrorCode()).isEqualTo(404)
        }
    }

    // @Test
    @Throws(Exception::class)
    fun event_with_future_date_does_not_fail() = runTest {
        login()
        val validEvent1 = EventUtils.createValidEvent()
        val validEvent2 = EventUtils.createEventWithFutureDate()
        val payload = EventPayload()
        payload.events = listOf(validEvent1, validEvent2)

        val response = postPayload(payload).getOrThrow()

        assertThat(response.response()!!.status()).isEqualTo(ImportStatus.SUCCESS)

        for (importSummary in response.response()!!.importSummaries()!!) {
            if (validEvent1.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            } else if (validEvent2.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            }
        }

        // Check server status
        val serverValidEvent1 = apiCallExecutor.wrap {
            eventService.getEvent(validEvent1.uid(), EventFields.allFields, ouMode)
        }.getOrThrow()

        val serverValidEvent2 = apiCallExecutor.wrap {
            eventService.getEvent(validEvent2.uid(), EventFields.allFields, ouMode)
        }.getOrThrow()

        assertThat(serverValidEvent1).isNotNull()
        assertThat(serverValidEvent2).isNotNull()
    }

    // @Test
    @Throws(Exception::class)
    fun event_with_invalid_program() = runTest {
        login()
        val validEvent = EventUtils.createValidEvent()
        val invalidEvent = EventUtils.createEventWithInvalidProgram()
        val payload = EventPayload()
        payload.events = listOf(validEvent, invalidEvent)

        val response = postPayload(payload).getOrThrow()

        assertThat(response.response()!!.status()).isEqualTo(ImportStatus.ERROR)

        for (importSummary in response.response()!!.importSummaries()!!) {
            if (validEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            } else if (invalidEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.ERROR)
            }
        }

        // Check server status
        val serverValidEvent = apiCallExecutor.wrap {
            eventService.getEvent(validEvent.uid(), EventFields.allFields, ouMode)
        }.getOrThrow()

        assertThat(serverValidEvent).isNotNull()

        try {
            apiCallExecutor.wrap {
                eventService.getEvent(
                    invalidEvent.uid(),
                    EventFields.allFields,
                    ouMode,
                )
            }.getOrThrow()
            Assert.fail("Should not reach that line")
        } catch (e: D2Error) {
            assertThat(e.httpErrorCode()).isEqualTo(404)
        }
    }

    // @Test
    @Throws(Exception::class)
    fun event_with_invalid_data_values() = runTest {
        login()
        val validEvent = EventUtils.createValidEvent()
        val invalidEvent = EventUtils.createEventWithInvalidDataValues()
        val payload = EventPayload()
        payload.events = listOf(validEvent, invalidEvent)

        val response = postPayload(payload).getOrThrow()

        assertThat(response.response()!!.status()).isEqualTo(ImportStatus.WARNING)

        for (importSummary in response.response()!!.importSummaries()!!) {
            if (validEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.SUCCESS)
            } else if (invalidEvent.uid() == importSummary.reference()) {
                EventUtils.assertEvent(importSummary, ImportStatus.WARNING)
            }
        }

        // Check server status
        val serverValidEvent = apiCallExecutor.wrap {
            eventService.getEvent(
                validEvent.uid(),
                EventFields.allFields,
                ouMode,
            )
        }.getOrThrow()

        val serverInvalidEvent = apiCallExecutor.wrap {
            eventService.getEvent(
                invalidEvent.uid(),
                EventFields.allFields,
                ouMode,
            )
        }.getOrThrow()

        assertThat(serverValidEvent).isNotNull()
        assertThat(serverValidEvent.trackedEntityDataValues()!!.size).isEqualTo(2)
        assertThat(serverInvalidEvent).isNotNull()
        assertThat(serverInvalidEvent.trackedEntityDataValues()).isNull()
    }

    private suspend fun postPayload(payload: EventPayload): Result<EventWebResponse, D2Error> {
        return apiCallExecutor.wrap(
            storeError = false,
            acceptedErrorCodes = listOf(409),
            errorClass = EventWebResponse::class.java,
        ) {
            eventService.postEvents(payload, strategy)
        }
    }

    private fun login() {
        d2.userModule().logIn(username, password, url, null).blockingGet()
    }
}
